server/src/internalClusterTest/java/org/opensearch/search/backpressure/SearchBackpressureIT.java

```diff
@@ -0,0 +1,312 @@
+        assertNotNull("SearchShardTask should have been cancelled with TaskCancelledException", caughtException);
+        assertEquals(TaskCancelledException.class, caughtException.getClass());
+        assertTrue(caughtException.getMessage().contains("cpu usage exceeded"));
+    }
+
+    public void testSearchShardTaskCancellationWithHighHeapUsage() throws InterruptedException {
+        // Before SearchBackpressureService cancels a task based on its heap usage, we need to build up the heap moving average
+        // To build up the heap moving average, we need to hit the same node with multiple requests and then hit the same node with a
+        // request having higher heap usage
+        String node = randomFrom(internalCluster().getNodeNames());
+        final int MOVING_AVERAGE_WINDOW_SIZE = 10;
+        Settings request = Settings.builder()
+            .put(SearchBackpressureSettings.SETTING_MODE.getKey(), "enforced")
+            .put(HeapUsageTracker.SETTING_HEAP_PERCENT_THRESHOLD.getKey(), 0.0)
+            .put(HeapUsageTracker.SETTING_HEAP_VARIANCE_THRESHOLD.getKey(), 1.0)
+            .put(HeapUsageTracker.SETTING_HEAP_MOVING_AVERAGE_WINDOW_SIZE.getKey(), MOVING_AVERAGE_WINDOW_SIZE)
+            .build();
+        assertAcked(client().admin().cluster().prepareUpdateSettings().setPersistentSettings(request).get());
+
+        CountDownLatch latch = new CountDownLatch(1);
+        ExceptionCatchingListener listener = new ExceptionCatchingListener(latch);
+        for (int i = 0; i < MOVING_AVERAGE_WINDOW_SIZE; i++) {
+            client(node).execute(TestTransportAction.ACTION, new TestRequest(RequestType.HIGH_HEAP), listener);
+        }
+
+        latch = new CountDownLatch(1);
+        listener = new ExceptionCatchingListener(latch);
+        client(node).execute(TestTransportAction.ACTION, new TestRequest(RequestType.HIGHER_HEAP), listener);
+        latch.await(TIMEOUT.getSeconds(), TimeUnit.SECONDS);
+
+        Exception caughtException = listener.getException();
+        assertNotNull("SearchShardTask should have been cancelled with TaskCancelledException", caughtException);
+        assertEquals(TaskCancelledException.class, caughtException.getClass());
+        assertTrue(caughtException.getMessage().contains("heap usage exceeded"));
+    }
+
+    public void testSearchCancellationWithBackpressureDisabled() throws InterruptedException {
+        Settings request = Settings.builder().put(SearchBackpressureSettings.SETTING_MODE.getKey(), "monitor_only").build();
+        assertAcked(client().admin().cluster().prepareUpdateSettings().setPersistentSettings(request).get());
+
+        CountDownLatch latch = new CountDownLatch(1);
+        ExceptionCatchingListener listener = new ExceptionCatchingListener(latch);
+        client().execute(TestTransportAction.ACTION, new TestRequest(RequestType.HIGH_ELAPSED_TIME), listener);
+        latch.await(TIMEOUT.getSeconds(), TimeUnit.SECONDS);
```

Comment:
When I run this test this call always returns false and the test takes the entire 30 seconds. Is that expected?

Reply:
Well, yes. We have set the TIMEOUT to be 30 seconds and it lets the request run for 30 seconds before timing out. We want to be sure that even if the search backpressure thresholds are breads, it is not cancelling the task.